[qilstiano] iP

qilstiano opened the pull request:
# 🧠 Einstein - A Genius Task Management Chatbot 🤖

> "Logic will get you from A to B. Imagination will take you everywhere." – Albert Einstein

Welcome to **Einstein**, your intelligent task management assistant! Einstein helps you manage your tasks efficiently with simple and intuitive commands.

---

## 🚀 Features

- 📝 **Add tasks**: Todo, Deadline, Event
- 📋 **List tasks**: View all tasks
- ✅ **Mark tasks**: Mark as done or not done
- 🗑️ **Delete tasks**: Remove tasks from the list
- 🔎 **Find tasks**: Search tasks by keyword
- 🔁 **Help**: View available commands
- 🚪 **Exit**: Close Einstein

---

## 📌 Commands

### 1. Add Tasks

#### 📝 Todo
```bash
todo <description>
```
Example:
```bash
todo read book
```

#### ⏰ Deadline
```bash
deadline <description> /by <date>
```
Example:
```bash
deadline return book /by Sunday
```

#### 🗓️ Event
```bash
event <description> /from <start> /to <end>
```
Example:
```bash
event project meeting /from Mon 2pm /to 4pm
```

---

## 🔍 Find Tasks
Search for tasks containing a keyword.

```bash
find <keyword>
```
Example:
```bash
find book
```

---

## 🛠️ Getting Started

1. **Download the JAR file** from [[here](https://github.com/qilstiano/ip/releases/tag/v0.1.1)].
2. **Run the application** using:
   ```bash
   java -jar Einstein.jar
   ```
3. **Start entering commands** to manage your tasks!

---

## ✅ Task List
- [x] Implement `find` command
- [x] Update `Parser` to handle `find`
- [ ] Improve search accuracy
- [ ] Simplify commands
- [ ] Add priorities
- [ ] Implement GUI for ease of use

---

## ⚠️ Error Handling
Einstein provides clear messages for errors:

```bash
> find 
```
```
Einstein
Oops! The 'find' command requires a keyword.
Usage: find <keyword>
```

---

## 🙏 Acknowledgments
Special thanks to:
- **Deepseek** for making the ASCII art and gradient text in the CLI

Enjoy using Einstein! 🚀

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

qilstiano committed:
feat/greeting + feat/farewell

Einstein
- added basic greeting
- added basic farewell

qilstiano committed:
Revert "feat/greeting + feat/farewell"

This reverts commit d3a0f150565d35709f871df317ed7f54b37ce54f.

qilstiano committed:
Create Einstein.class

qilstiano committed:
Revert "Create Einstein.class"

This reverts commit ade6b78b5a23f26cf260adca11b04a4a2c3a7545.

qilstiano committed:
feat/einstein

feat/greeting
feat/farewell

qilstiano committed:
feat/echo

feat/echo
styles/ascii

qilstiano committed:
feat/tasks

feat/add-tasks
feat/list-tasks

qilstiano committed:
feat/tasks

feat/mark-done
feat/unmark-done
styles/gradient-ascii

qilstiano committed:
feat/tasks

feat/todo
feat/event
feat/deadline

qilstiano committed:
testing/automated-ui-test

qilstiano committed:
feat/error-handling

qilstiano committed:
feat/delete

- feat/delete
- feat/help
- styles/einstein-response
- styles/responder

qilstiano committed:
README/v1

basic README added

qilstiano committed:
update README.md

qilstiano committed:
Add task persistence with file-based storage

Implement automatic saving and loading of tasks to/from a file to ensure data persistence.

* Save tasks to `./data/duke.txt` whenever the task list changes.
* Load tasks from the file when the chatbot starts.
* Handle cases where the file or directory does not exist (Stretch goal)
* Add basic handling for corrupted data in the file.
* Use OS-independent file paths for compatibility across systems.

This ensures that tasks are preserved between sessions and improves the chatbot's usability.

qilstiano committed:
Merge branch 'branch-Level-7'

Add task persistence with file-based storage

- Save tasks to `./data/duke.txt` whenever the task list changes.
- Load tasks from the file when the chatbot starts.
- Handle cases where the file or directory does not exist.
- Add basic handling for corrupted data in the file.
- Use OS-independent file paths for compatibility across systems.

This ensures that tasks are preserved between sessions and improves the chatbot's usability.

qilstiano committed:
Update Event and Deadline classes to use LocalDateTime

Refactor Event and Deadline classes to use java.time.LocalDateTime for date and time handling.

* Replace String-based date fields with LocalDateTime in Event and Deadline classes.
* Add date parsing logic to handle user input in formats like "dd/MM/yyyy HHmm".
* Format dates for display in a user-friendly format (e.g., "MMM dd yyyy, h:mm a").
* Update file saving and loading logic to support LocalDateTime serialization.

This change improves date and time handling, making it more robust and user-friendly.

qilstiano committed:
Add date parsing, update deadline/event handling, and implement stretch goal

Enhance chatbot to handle dates and times effectively.

* Add date parsing logic using `DateTimeFormatter` to parse user input (e.g., "dd/MM/yyyy HHmm").
* Update `Deadline` and `Event` classes to use `LocalDateTime` for date and time storage.
* Modify `handleDeadlineCommand` and `handleEventCommand` to parse and store dates correctly.
* Implement stretch goal: Add `list <date>` command to display tasks occurring on a specific date.
* Update file saving and loading logic to support `LocalDateTime` serialization.

This change improves the chatbot's ability to understand and work with dates and times, providing a more intuitive user experience.

qilstiano committed:
Update file handling and help function

Enhance chatbot functionality and improve file handling for date and time support.

* Update `displayHelp` function to include the new `list <date>` command and reflect the updated date format for deadlines and events.
* Modify `saveTasksToFile` and `loadTasksFromFile` to handle `LocalDateTime` .
* Ensure file saving and loading supports the new date and time format (dd/MM/yyyy HHmm).

These changes improve user guidance and ensure consistent handling of dates across file operations.

qilstiano committed:
Update processCommands for date and time functionality

Enhance processCommands to support new date and time features.

* Add support for parsing and handling dates in `deadline` and `event` commands.
* Implement the `list <date>` command to display tasks occurring on a specific date.
* Improve error handling for invalid date formats and commands.
* Ensure seamless integration with the updated `Deadline` and `Event` classes.

This change enables the chatbot to understand and process dates and times effectively, improving user interaction.

qilstiano committed:
Refactor list command handling to encapsulate logic in listTasks

Encapsulate the logic for handling the list command within the listTasks method.

* Modify listTasks to handle both "list" and "list <date>" commands.
* Remove specific handling of "list <date>" from processCommands.
* Simplify processCommands to only call listTasks for list-related commands.
* Ensure listTasksByDate remains unchanged and is called internally by listTasks.

This refactoring improves code readability and maintainability by centralizing the logic for listing tasks.

qilstiano committed:
Integrate branch-Level-8 into master, adding support for date and time handling.

Key changes included in this merge:
* Added date parsing and formatting using `java.time.LocalDateTime`.
* Updated `Deadline` and `Event` classes to store dates as `LocalDateTime`.
* Implemented the `list <date>` command to filter tasks by a specific date.
* Refactored `listTasks` to handle both "list" and "list <date>" commands.
* Improved error handling for invalid date formats and commands.
* Updated file saving and loading to support `LocalDateTime` serialization.

This merge enhances the chatbot's functionality by enabling robust date and time handling, improving user interaction and task management.

qilstiano committed:
Refactor: Apply OOP principles across classes and methods

Refactoring of methods and classes to improve modularity, reusability, and maintainability of the codebase.

By applying OOP principles, the code becomes easier to extend and maintain for better long-term scalability.

qilstiano committed:
Refactor codebase to adhere to OOP principles

Refactor the codebase to improve modularity, maintainability, and adherence to OOP principles.

Key changes:
- Extracted closely related code into separate classes:
  - `Ui`: Handles all user interactions.
  - `Storage`: Manages loading and saving tasks to/from a file.
  - `Parser`: Parses user commands and returns appropriate `Command` objects.
  - `TaskList`: Manages the list of tasks (e.g., adding, deleting, marking tasks).
- Introduced the `Command` interface and implemented command classes (e.g., `AddCommand`, `DeleteCommand`, `ExitCommand`) to encapsulate the logic for each user command.
- Simplified the `Einstein` class to delegate responsibilities to the extracted classes, improving readability and maintainability.

Benefits of this refactoring:
1. **Modularity**: Each class has a single responsibility, making the code easier to understand and extend.
2. **Reusability**: Common functionality (e.g., parsing dates, saving tasks) is centralized and reusable.
3. **Maintainability**: Changes to one part of the code (e.g., command logic) are isolated and less likely to affect other parts.
4. **Scalability**: Adding new commands or features is now straightforward, as the structure is well-organized.

This refactoring ensures the codebase is more robust, flexible, and aligned with OOP best practices.

qilstiano committed:
Merge branch 'branch-A-Packages'

Refactor the codebase to organize classes into suitable Java packages.

Key changes:
- Created the following packages under `duke`:
  - `einstein.command`: Contains all command-related classes (e.g., `AddCommand`, `DeleteCommand`).
  - `einstein.exception`: Contains custom exceptions (e.g., `EinsteinException`).
  - `einstein.storage`: Contains file storage-related classes (e.g., `Storage`).
  - `einstein.task`: Contains task-related classes (e.g., `Task`, `Todo`, `Deadline`, `Event`).
  - `einstein.ui`: Contains user interaction-related classes (e.g., `Ui`).
- Updated package declarations and import statements in all classes to reflect the new package structure.
- Moved the main application class (`Einstein`) to the root `einstein` package.

Benefits of this refactoring:
1. **Modularity**: Classes are grouped into packages based on their responsibilities, making the codebase easier to navigate and maintain.
2. **Readability**: The package structure provides a clear separation of concerns, improving code readability.
3. **Scalability**: The organized structure makes it easier to add new features or classes in the future.

This change aligns with OOP principles and improves the overall structure of the codebase.

qilstiano committed:
Merge remote-tracking branch 'origin/add-gradle-support'

qilstiano committed:
Add unit tests for Parser, TaskList, and Storage

Added JUnit tests for Parser, TaskList, and Storage to ensure correct behavior.

- ParserTest: Verifies that parsing "todo" and "deadline" commands returns the correct command objects and handles invalid input properly.
- StorageTest: Ensures tasks are saved to and loaded from a file correctly.
- TaskListTest: Checks adding and deleting tasks, including handling invalid deletions.

These tests improve code reliability and facilitate future refactoring.

Let's:
- Continue to regularly update test cases and implement future test cases for future functionalities (stretch goal)

qilstiano committed:
Update build.gradle

yuqing-tham commented on src/main/java/einstein/Einstein.java:
```diff
@@ -0,0 +1,46 @@
+
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
Good job on naming the boolean `isExit` correctly!

yuqing-tham commented on src/main/java/einstein/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+
+public class AddDeadlineCommand implements Command {
+    private String description;
+    private LocalDateTime by;
```
Perhaps a more descriptive name such as `deadline` or `dueDate`?

qilstiano replied:
That sounds like a great suggestion. Thank you! I'll include it in my future commits 👍

yuqing-tham commented on src/main/java/einstein/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+    private LocalDateTime by;
+
+    public AddDeadlineCommand(String fullCommand) throws EinsteinException {
+        String[] parts = fullCommand.substring(9).split("/by", 2);
```
Good job on the use of plural forms for an array.

yuqing-tham commented on src/main/java/einstein/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+        this.by = parseDateTime(parts[1].trim());
+    }
+
+    private LocalDateTime parseDateTime(String dateTimeStr) throws EinsteinException {
```
Kudos on using a verb to name a method performing an action.

yuqing-tham commented on src/main/java/einstein/command/AddEventCommand.java:
```diff
@@ -0,0 +1,51 @@
+package einstein.command;
+
+// Imports for Date and Time
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import einstein.exception.EinsteinException;
+import einstein.task.Task;
+import einstein.task.Event;
+import einstein.storage.TaskList;
+import einstein.storage.Storage;
+import einstein.ui.Ui;
```
I like how the imports are in a very consistent order.

yuqing-tham commented on src/main/java/einstein/storage/Storage.java:
```diff
@@ -0,0 +1,108 @@
+import einstein.task.Event;
+import einstein.task.Deadline;
+import einstein.exception.EinsteinException;
+
```
Perhaps you can add header comments for public classes and methods in your code?

qilstiano replied:
Great suggestion! I've added a couple of them for some methods and classes today. Let me know if they look good if you get the chance to check it out!

yuqing-tham commented on src/main/java/einstein/storage/Storage.java:
```diff
@@ -0,0 +1,108 @@
+              String description = parts[2].trim();
+
+              Task task;
+              switch (type) {
+                  case "T":
```
Please check the coding standard and change the settings in your IDE accordingly!

qilstiano replied:
Thanks for spotting that! Must have forgot to fix this switch statement after I changed my settings. Really appreciate you flagging it out 👍

yuqing-tham commented on src/main/java/einstein/storage/TaskList.java:
```diff
@@ -0,0 +1,51 @@
+        tasks.remove(index);
+    }
+
+    public void markTaskAsDone(int index) throws EinsteinException {
```
Naming of the method is descriptive and representative of the actions being performed, good job!

yuqing-tham reviewed (commented):
LGTM overall, only some minor fixes needed. Keep up the good work.

qilstiano committed:
Add Javadocs to Parser class

qilstiano committed:
Add Javadocs for EinsteinException

qilstiano committed:
Add Javadocs for Storage and TaskList

Added Javadocs for code readability
Minor edits to EinsteinException and
Storage code style

qilstiano committed:
Add Javadocs for Deadline, Event, Task and Todo

- Add Javadocs for code readability
- Rectify code style issues

nicolaskjh commented on src/main/java/einstein/ui/Ui.java:
```diff
@@ -0,0 +1,106 @@
+      }
+  }
+
+  public void showTasksByDate(ArrayList<Task> tasks, LocalDate date) {
```
Good job on using verbs and having descriptive method names!

nicolaskjh commented on src/main/java/einstein/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+import einstein.task.Deadline;
+import einstein.storage.TaskList;
+import einstein.storage.Storage;
+import einstein.ui.Ui;
```
Good job on importing the required classes explicitly and keeping the order consistent!

nicolaskjh commented on src/main/java/einstein/command/MarkCommand.java:
```diff
@@ -0,0 +1,31 @@
+import einstein.storage.Storage;
+
+public class MarkCommand implements Command {
+    private int taskIndex;
```
Good job on having an intuitive variable name for this attribute

nicolaskjh commented on src/main/java/einstein/storage/Storage.java:
```diff
@@ -0,0 +1,108 @@
+          DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+
+          for (Task task : tasks) {
+              if (task instanceof Todo) {
+                  data.append("T | ").append(task.isDone ? "1" : "0").append(" | ").append(task.description).append("\n");
+              } else if (task instanceof Deadline) {
+                  Deadline deadline = (Deadline) task;
+                  data.append("D | ").append(deadline.isDone ? "1" : "0").append(" | ").append(deadline.description)
+                      .append(" | ").append(deadline.by.format(formatter)).append("\n");
+              } else if (task instanceof Event) {
+                  Event event = (Event) task;
+                  data.append("E | ").append(event.isDone ? "1" : "0").append(" | ").append(event.description)
+                      .append(" | ").append(event.from.format(formatter)).append(" | ").append(event.to.format(formatter)).append("\n");
+              }
```
Perhaps enums could be used here to replace instanceof?

qilstiano replied:
That's a great suggestion! I'll keep that in mind for my future commits.

nicolaskjh commented on src/test/java/einstein/storage/StorageTest.java:
```diff
@@ -0,0 +1,44 @@
+class StorageTest {
+
+    @Test
+    void saveAndLoadTasks_validTasks_tasksSavedAndLoadedSuccessfully() throws EinsteinException, IOException {
```
Good job on following the naming conventions for tests!

nicolaskjh reviewed (commented):
Good job so far! Keep up the good work!

qilstiano committed:
Add SearchCommand to support task searching

Currently, users have no way to search for tasks by keyword in their
descriptions. This makes it difficult to quickly locate specific tasks
in long task lists.

Introduce `FindCommand`, which allows users to retrieve tasks
containing a specified keyword. Additionally, update `Parser`
to recognize the `find` command and invoke `SearchCommand` accordingly.

With this enhancement, users can efficiently filter their task list
using a simple search query.

qilstiano committed:
Refactor: Ensure code adheres to Coding Standard

Current codebase contains inconsistencies and deviations from our
established Coding Standard. This affects code readability and
maintainability.

Let's update the code to strictly follow our Coding Standard:

• Correct indentation and formatting across all files
• Standardize naming conventions for variables and functions
• Remove unused imports and variables
• Add necessary comments for complex logic
• Reorganize code structure for better modularity

This refactoring will improve code quality, make future maintenance
easier, and help new developers understand the codebase more quickly.

qilstiano committed:
Merge: Add JavaDoc to Einstein chatbot

The Einstein chatbot currently lacks comprehensive documentation,
making it difficult for developers to understand and maintain the code.

This merge adds JavaDoc comments to the Einstein chatbot codebase:

• Document public methods and classes
• Add package-level documentation
• Include usage examples for complex algorithms
• Clarify parameter descriptions and return values

These changes will improve code readability, facilitate easier
maintenance, and help new developers onboard more quickly.

qilstiano committed:
Merge: Align Einstein project with CS2103 Coding Standard

The Einstein project currently deviates from the CS2103 Coding Standard,
which can lead to inconsistencies and reduced code readability.

This merge updates the codebase to adhere to the current CS2103 Coding
Standard:

• Standardize naming conventions for variables, methods, and classes
• Adjust indentation and formatting across all files
• Reorganize code structure for improved modularity
• Remove unused imports and variables
• Add necessary comments for complex logic

These changes will enhance code quality, facilitate easier maintenance,
and improve collaboration among team members.erge branch 'branch-A-CodingStandard'

qilstiano committed:
Merge: Add task search and display feature to Einstein

Einstein currently lacks the ability to find and show user tasks based
on queries, limiting its usefulness as a personal assistant.

This merge introduces a new feature that allows Einstein to:

• Parse user queries for task-related keywords
• Search through the user's task database efficiently
• Display relevant tasks matching the query criteria
• Format task information for clear presentation

These enhancements will significantly improve Einstein's functionality
as a task management tool, making it more valuable to users.

qilstiano committed:
Update version in ASCII

qilstiano committed:
Update build.gradle for GUI

qilstiano committed:
Add DialogBox, Main, MainWindow and Images

qilstiano committed:
Refactor Duke to use FXML for GUI

Duke's GUI code is currently intertwined, making it harder to maintain and extend.

Using FXML allows separation of the GUI design (view) from the code that controls it (controller), improving maintainability and extensibility.

Let's:
* Extract a `MainWindow` class to contain the controller logic for the main GUI.
* Create `MainWindow.fxml` in `src/main/resources/view/` to define the view for the main window.
* Create `DialogBox.fxml` in `src/main/resources/view/` to define the view for the dialog boxes.
* Trim down `DialogBox.java` to only contain the controller logic.

This separation of concerns will make the codebase more modular and easier to work with.

qilstiano committed:
Add basic UI styling to GUI

qilstiano committed:
Fix all Commands to work with GUI

Commands were previously intended to work with CLI by using
System.out.println commands.

With this push, we are now able to receive responses by Einstein
on the GUI.

qilstiano committed:
Merge branch-Level-10 for JavaFX GUI support

This merge incorporates a JavaFX-based GUI for Einstein,
enhancing user interaction with a more intuitive and visually
appealing interface.

Changes:
* Implemented JavaFX GUI with basic layout and classic grey and orange
Einstein styling.
* Update Ui methods to return formatted strings instead of
  'System.out.println' calls.
* Modify Command classes to return strings in 'execute'.
* Ensure compatibility between CLI and GUI modes.

qilstiano committed:
Add Checkstyle support to Einstein

qilstiano committed:
Add Checkstyle support to Einstein

This merge introduces Checkstyle to Einstein to
ensure better adherence to code style guidelines.

The following changes were made in the merged branch:

* Added 'checkstyle.xml' and 'suppressions.xml'
to config. These files contain the code style
and suppression configurations respectively.
* Updated the 'build.gradle' file to include
the Checkstyle plugin (v10.2)
* Configured Checkstyle to run during build
processes.

These changes will help maintain a consistent
code style throughout the project, improving
readability and maintainability.

qilstiano committed:
Fix code style & class encapsulation

Task class description and isDone fields are currently public.

This violates encapsulation principles and exposes internal state
directly, making the class harder to maintain and prone
to unintended modifications. Additionally, code style was not
up to mark.

With this commit, the key changes are:
*   Make the description and isDone fields private.
*   Add getter methods for retrieving the values of these fields.
*   Add setter methods for modifying the values of these fields.
*  Amended all code style errors with checkStyle plugin

This improves encapsulation, provides controlled access to the
task's state, and increases the class's robustness. This commit also
fixes glaring code style issues to align with CS2103 coding
style standards for greater readability.

qilstiano committed:
Minor response changes

qilstiano committed:
Add support for varargs in Todo command

The 'todo' command now allows users to add
multiple todos in a single command.

qilstiano committed:
Merge to support multiple todos

The 'todo' command now allows users to add multiple
todos in a single command. This enhances the user
experience by reducing the need for multiple commands
when adding several todos. Additionally, various
code style fixes and class encapsulation changes
were made to enhance code readability and quality.

Changes:
* Updated 'Parser' to split input into multiple
todo descriptions.
* Created 'AddMultipleTodosCommand' to handle
addition of multiple todos using varargs.
* Modified 'AddTodoCommand' to ensure it processes
a single task description.
* Fixed code style issues using checkStyle plugin
* Fixed code quality issues in class encapsulation

To do:
* Implement varargs for delete command to allow for
convenient batch deleting of Tasks.

qilstiano committed:
Remove Einstein.class file from tracking

qilstiano committed:
Add assertions for all Add* Commands

qilstiano committed:
Add assertions to remaining Command classes

The Command classes lacked explicit checks for critical
assumptions, potentially leading to unexpected behaviour
and difficulty debugging.

Adding assertions improves code robustness by validating
preconditions, postconditions and invariants at runtime,
catching potential errors early.

Let's:
* Add assertions to:
	- DeleteCommand
	- ExitCommand
	- FindCommand
	- HelpCommand
	- ListByDateCommand
	- ListCommand
	- MarkCommand
	- UnmarkCommand
* Add support to close Einstein in the ExitCommand
using JavaFX.

Using assertions provides a clear and concise way
to document assumptions and improve the overall
reliability of the changed classes. Additionally,
fixing the ExitCommand post-GUI introduction
ensures users are able to exit Einstein via chatbot
input.

To-do:
* Add  assertions to remainder of classes,
where relevant.

qilstiano committed:
Add assertions to remaining Command classes

The Command classes lacked explicit checks for critical
assumptions, potentially leading to unexpected behaviour
and difficulty debugging.

Adding assertions improves code robustness by validating
preconditions, postconditions and invariants at runtime,
catching potential errors early.

Key changes:
* Added assertions to:
	- DeleteCommand
	- ExitCommand
	- FindCommand
	- HelpCommand
	- ListByDateCommand
	- ListCommand
	- MarkCommand
	- UnmarkCommand
* Added support to close Einstein in the ExitCommand
using JavaFX.

Let's:
* Continue adding assertions to remainder
of classes where relevant.

qilstiano committed:
Merge branch 'branch-A-Assertions' of https://github.com/qilstiano/ip into branch-A-Assertions

qilstiano committed:
Merge branch 'branch-A-Assertions' of https://github.com/qilstiano/ip into branch-A-Assertions

This merge was done due to commit amendments to the previous commit.

qilstiano committed:
Merge branch 'branch-A-Assertions' of https://github.com/qilstiano/ip into branch-A-Assertions

qilstiano committed:
Enhance code reliability with assertions across core classes

Several core classes in the Einstein project lack explicit checks for critical assumptions, potentially leading to unexpected behavior and difficult debugging.

Adding assertions improves code robustness by validating preconditions, postconditions, and invariants at runtime, catching potential errors early in development. This proactive approach reduces the risk of subtle bugs and improves the overall reliability of the system.

Let's add assertions to the Deadline, Einstein, Event, Parser, Storage, Task, Todo, and Ui classes to verify:
- Non-null and valid input parameters.
- Expected object states at key points in the code.

Using assertions provides a clear and concise way to document assumptions and improve the overall reliability of the core classes. This will make future debugging easier and less time-consuming.

qilstiano committed:
Refactor AddDeadlineCommand, AddEventCommand, and Storage classes

- Apply Single Level of Abstraction Principle (SLAP)
- Break down long methods into smaller, focused ones
- Reduce code duplication across classes
- Improve overall code quality and maintainability

This commit improves the structure and readability
of the codebase by adhering to best
practices in code organization and design.

qilstiano committed:
Revert "Refactor AddDeadlineCommand, AddEventCommand, and Storage classes"

This reverts commit 4a13ed4ccdae5787c7c113ec1301514e8dc91884.

qilstiano committed:
Refactor AddDeadlineCommand, AddEventCommand, and Storage classes

- Apply Single Level of Abstraction Principle (SLAP)
- Break down long methods into smaller, focused ones
- Reduce code duplication across classes
- Improve overall code quality and maintainability

This commit improves the structure and readability of the codebase
by adhering to best practices in code organization and design.

qilstiano committed:
Merge pull request #2 from qilstiano/branch-A-Assertions

Introduce Assertions to improve debugging

qilstiano committed:
Merge branch 'master' into branch-A-CodeQuality

qilstiano committed:
Refactor to improve code quality and readability

Previously, some Einstein classes lacked readability as some methods
were extremely long or had bad nesting. 

The AddDeadlineCommand, AddEventCommand and Storage classes 
have been refactored to improve code quality and maintainability. 

Let's:
*   Adhere to the Single Level of Abstraction Principle (SLAP) by breaking down long methods into smaller, focused ones.
*   Reduce code duplication by extracting common logic into reusable methods.
*   Add assertions to ensure data integrity and robustness.
*   Improve overall code structure and readability.

These changes enhance the maintainability and reliability of the changed class, making it easier to understand and extend in the future. This overall improves
readability for developers.

qilstiano committed:
Add support for Github Actions CI

qilstiano committed:
Add support for tags in Task

qilstiano committed:
Merge branch 'branch-A-CodingStandard'

qilstiano committed:
Add tagging functionality to task management system

Tasks do not support tagging, limiting the ability to categorize
and filter them efficiently.

Users need a way to associate tags with tasks for better organization and retrieval.

Let's:
- Add `tags` field (Set<String>) to `Task` class for storing associated tags.
- Implement `addTag(String tag)`, `removeTag(String tag)`,
and `getTags()` methods in `Task` class for tag management.
- Update `toString()` method in `Task` class to include tags in task representation.
- Modify `Parser` class to recognize `tag` and `untag` commands.
- Extend `parse()` method to create `AddTagCommand` and `RemoveTagCommand` objects.
- Create `AddTagCommand` to handle tag addition by
extracting task index and tag.
- Create `RemoveTagCommand` to handle tag removal by
extracting task index and tag.
- Update `Ui` class with `showTagAdded(Task task, String tag)`
and `showTagRemoved(Task task, String tag)` methods.

Using a `Set<String>` ensures uniqueness and efficient retrieval of tags.
Implementing separate command classes follows the existing command-based
architecture, maintaining consistency and modularity.

This update enhances task management by allowing users to tag and untag tasks.

qilstiano committed:
Tidy up code and amend bugs

Key changes:
- Fix bug where tag which does not exist somehow gets removed
by 'RemoveTagCommand' - amended to reflect error message
- Fix bug where welcome message does not show to user upon
initialize - amended in showWelcome() in Ui
- Minor code quality and code style amendments
- Added label to JavaFX MainWindow

qilstiano committed:
Add GUI improvements

The GUI of Einstein previously lacked a consistent look and feel.
Hence, a small overhaul was needed to give the UI some consistency.

Let's:
- Remove the background from the DaUser and DaEinstein images
- Give the dialogue bubble for Einstein a different colour from User.
- Amend codestyle errors project-wide

With these changes, Einstein's GUI now has
a more consistent look and feel, which can help enhance
UX.

qilstiano committed:
Update README.md

qilstiano committed:
Update README.md

qilstiano committed:
Update README.md

qilstiano committed:
Create Priority Enums for task priority

Einstein currently has no way to represent task priority
(unless user uses tags). This may lead to inconsistencies
and makes it difficult to manage or update priorities in
the future.

Let's:
- Introduce an 'Enums' class with an enum called
'Priority' to define 4 priority levels for Einstein.

Using an 'enum' provides type safety, and improves overall
code reliability and quality. This also simplifies future
modifications to priority options.

qilstiano committed:
Merge branch 'master' of https://github.com/qilstiano/ip

qilstiano committed:
Attach revised DaUser.png

qilstiano committed:
Attach Ui.png for submission

qilstiano committed:
Update README.md

qilstiano committed:
Update README.md

qilstiano committed:
Update README.md

qilstiano committed:
Amend build.gradle file to output for Java 17

qilstiano committed:
Merge branch 'master' of https://github.com/qilstiano/ip

qilstiano committed:
Update README.md with project details

README.md lacks project documentation  

Users currently have no reference for Einstein’s features, usage, or setup. This makes it difficult to understand how to interact with the chatbot.  

Add a detailed README.md explaining Einstein’s functionality, commands, setup instructions, and error handling.  

Providing clear documentation ensures users can easily set up and use Einstein. The structured format improves readability and accessibility.  

Includes usage examples, error messages, and acknowledgments for better user experience.

qilstiano committed:
Update README.md

Minor typo

qilstiano committed:
Add Priority feature to Task

Tasks currently do not have a priority level, making it difficult for users to prioritize their tasks effectively.

Adding a priority feature will allow users to assign priority levels (UBER_HIGH, HIGH, MEDIUM, LOW) to tasks, helping them focus on what matters most.

Let's:

* Add a Priority enum to represent priority levels (HIGH, MEDIUM, LOW).

* Add a priority field to the Task class.

* Add methods to set and get the priority in the Task class.

* Create a SetPriorityCommand to handle setting the priority for a task.

* Update the Parser class to recognize the priority command.

* Update the Ui class to display a message when the priority is set.

This implementation ensures that users can easily assign and manage
task priorities, improving task management efficiency. The use of enums ensures
type safety and clarity in priority levels.

qilstiano committed:
Merge remote-tracking branch 'origin/master'

qilstiano committed:
Amend README.md and HelpCommand to account for Priority

The user guide(UG)/README.md lacked documentation on the newly-introduced Priority feature for Tasks. This is
also the case for the HelpCommand.

This needs to change to ensure that the UG and HelpCommand is comprehensive and that new users are able to be
quickly onboarded by simply reading the UG.

Let's
- Append details regarding the implementation of the Priority system
- Provide details on how users can use the Priority system including syntax and potential errors
- Append similar details to HelpCommand.java

By doing so, this gives users enough clarity when reading the UG or
using the HelpCommand via 'help'.

qilstiano committed:
Add Assertions to SetPriorityCommand

The SetPriorityCommand class lacks internal checks to ensure data integrity,
potentially leading to unexpected behavior and difficult debugging.

Adding assertions enhances the class's robustness by validating preconditions and
postconditions, catching errors early and providing clearer failure points.

Let's:
- Add assertions to the constructor to validate input parameters (fullCommand,
  taskIndex, priority).
- Add assertions to the execute method to validate TaskList, Ui, Storage, and
  the state of tasks.

These assertions ensure:
- Input parameters are not null or empty where appropriate.
- Task index is within valid bounds.
- Required objects (TaskList, Ui, Storage) are not null.
- The task at the specified index exists.
- Result string is not null

This approach provides more confidence in the correctness of the class and
simplifies debugging by pinpointing the exact location of failures.

qilstiano committed:
Add assertions to RemoveTagCommand for increased validation

The RemoveTagCommand class currently lacks comprehensive input validation and
object state checks, which can lead to runtime errors and hinder debugging.

Adding assertions enhances the class's reliability by validating inputs and
ensuring object integrity at critical points in the execution.

Let's:
- Add assertions to the constructor to validate fullCommand, taskIndex, and tag.
  Ensure they are not null where appropriate and taskIndex is valid.
- Add assertions to the execute method to validate TaskList, Ui, and Storage.
  Ensure they are not null, and validate the task list and task at the given
  index.

These validations help:
- Catch invalid input early, preventing unexpected behavior.
- Ensure the TaskList, Ui, and Storage dependencies are correctly injected.
- Confirm the task exists at the specified index before attempting to remove
  the tag.
- Validate the result to confirm the expected message to the user

This validation strategy increases confidence in the class's
correctness and simplifies debugging by pinpointing the source of errors.

qilstiano committed:
Add assertions to AddTagCommand for input and state validation

The AddTagCommand class is currently missing checks for input validity
and object state, potentially leading to unexpected behavior
or errors during runtime.

Adding assertions improves the class's reliability by validating command
parameters and ensuring the TaskList, Ui, and Storage objects are in a
consistent state.

Let's:
- Add assertions to the constructor to validate fullCommand, taskIndex, and tag.
  Ensure they are non-null and non-empty where appropriate and taskIndex is
  valid.
- Add assertions to the execute method to validate TaskList, Ui, and Storage.
  Ensure they are not null, and validate the task list and task at the given
  index are also not null.
- Validate the result string returned after adding the tag to ensure there are no unexpected issues

These validations:
- Catch invalid inputs early, preventing potential runtime errors.
- Ensure essential dependencies (TaskList, Ui, Storage) are correctly
  provided.
- Confirm the task at the given index exists before attempting to add the tag.
- Validate the result string to confirm to user it worked

This proactive approach enhances the class's robustness and makes debugging
easier by clearly identifying the source of errors.

qilstiano committed:
Add Javadoc comments to test classes for better documentation

The test classes (ParserTest, StorageTest, TaskListTest) currently lack
Javadoc comments, making it harder to understand the purpose and
functionality of each test case.

Adding Javadoc comments improves the readability and maintainability of
the test suite by providing clear documentation for each class and test
method.

Let's:
- Add class-level Javadoc comments to ParserTest, StorageTest, and
  TaskListTest, explaining the purpose of each class.
- Add method-level Javadoc comments to each test method, describing what
  the test is checking, what it's verifying, and any exceptions that
  might be thrown.

These comments will:
- Make it easier for developers to understand the intent and behavior
  of each test.
- Serve as a valuable reference for future maintenance and modifications
  of the test suite.
- Improve the overall quality and clarity of the codebase.

This enhancement helps ensure the tests are well-documented, contributing to
the long-term maintainability of the project.

qilstiano committed:
Add AI.md to document AI-assisted contributions

Add AI.md to outline the specific ways AI was used in the development of the Einstein chatbot.
This includes AI-assisted documentation, README formatting, debugging support, and assertion writing.
The document also clarifies human oversight in AI-generated content.

This improves project transparency regarding AI usage.

qilstiano committed:
Update README.md
